### PR TITLE
Viktor lab1

### PR DESCRIPTION
**Pull request by bdgToji**
FTR: Implemented the first laboratory exercise

### CODE REVIEW

File: Лабораториска 1/PipeAndFilter/src/main/java/PipeAndFilterProblem.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        ClassLoader loader = PipeAndFilterProblem.class.getClassLoader();
+        Scanner scanner = new Scanner(new File(loader.getResource("student_grades.csv").getFile()));
+
+        Pipe<String> pipe1 = new Pipe<>();
+        Pipe<String> pipe2 = new Pipe<>();
+
+        ToLowerFilter toLowerFilter = new ToLowerFilter();
+        NoNamesFilter noNamesFilter = new NoNamesFilter();
+        GroupCoursesFilter groupCoursesFilter = new GroupCoursesFilter();
+        GradeFilter gradeFilter = new GradeFilter();
```
Comment (vasilaking) [inline Лабораториска 1/PipeAndFilter/src/main/java/PipeAndFilterProblem.java:26]:
```suggestion
```

File: Лабораториска 1/PipeAndFilter/src/main/java/PipeAndFilterProblem.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        pipe1.addFilter(toLowerFilter);
+        pipe1.addFilter(noNamesFilter);
+        pipe1.addFilter(groupCoursesFilter);
+
```
Comment (vasilaking) [inline Лабораториска 1/PipeAndFilter/src/main/java/PipeAndFilterProblem.java:30]:
```suggestion

        GradeFilter gradeFilter = new GradeFilter();
```

File: Лабораториска 1/PipeAndFilter/src/main/java/PipeAndFilterProblem.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+                    numberPassing.put(fields[1], numberPassing.get(fields[1]) + 1);
+                }
+                else {
+                    numberPassing.put(fields[1], 1.0);
+                }
+
+                //System.out.println(line1);
+            }
+
+            System.out.println(line1);
```
Comment (vasilaking) [inline Лабораториска 1/PipeAndFilter/src/main/java/PipeAndFilterProblem.java:63]: Why is "System.out.println(line1);" changed from inside the inner if to the outer if?

File: Лабораториска 1/PipeAndFilter/src/main/java/service/impl/GradeFilter.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
+package service.impl;
+
+import service.Filter;
+
+import java.util.Arrays;
+
+public class GradeFilter implements Filter {
```
Comment (vasilaking) [inline Лабораториска 1/PipeAndFilter/src/main/java/service/impl/GradeFilter.java:7]:
```suggestion
public class GradeFilter<T> implements Filter<String> {
```

File: Лабораториска 1/PipeAndFilter/src/main/java/service/impl/GradeFilter.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    @Override
+    public String execute(Object input) {
```
Comment (vasilaking) [inline Лабораториска 1/PipeAndFilter/src/main/java/service/impl/GradeFilter.java:9]:
```suggestion
    public String execute(String input) {
```

File: Лабораториска 1/PipeAndFilter/src/main/java/service/impl/GroupCoursesFilter.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
+package service.impl;
+
+import service.Filter;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+
+public class GroupCoursesFilter implements Filter {
```
Comment (vasilaking) [inline Лабораториска 1/PipeAndFilter/src/main/java/service/impl/GroupCoursesFilter.java:9]:
```suggestion
public class GroupCoursesFilter<T> implements Filter<String> {
```

File: Лабораториска 1/PipeAndFilter/src/main/java/service/impl/GroupCoursesFilter.java [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+        courseGroups.put("discrete mathematics", "math");
+        courseGroups.put("probability and statistics", "math");
+        courseGroups.put("software architecture and design", "software engineering");
+        courseGroups.put("software requirements analysis", "software engineering");
+        courseGroups.put("structural programming", "programming");
+        courseGroups.put("object oriented programming", "programming");
+        courseGroups.put("algorithms and data structures", "programming");
+    }
+    @Override
+    public String execute(Object input) {
```
Comment (vasilaking) [inline Лабораториска 1/PipeAndFilter/src/main/java/service/impl/GroupCoursesFilter.java:29]:
```suggestion
    public String execute(String input) {
```

File: Лабораториска 1/PipeAndFilter/src/main/java/service/impl/NoNamesFilter.java [hunk 8]
```diff
@@ -0,0 +1,20 @@
+package service.impl;
+
+import service.Filter;
+
+import java.util.Arrays;
+
+public class NoNamesFilter implements Filter {
```
Comment (vasilaking) [inline Лабораториска 1/PipeAndFilter/src/main/java/service/impl/NoNamesFilter.java:7]:
```suggestion
public class NoNamesFilter<T> implements Filter<String> {
```

File: Лабораториска 1/PipeAndFilter/src/main/java/service/impl/NoNamesFilter.java [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+    @Override
+    public String execute(Object input) {
```
Comment (vasilaking) [inline Лабораториска 1/PipeAndFilter/src/main/java/service/impl/NoNamesFilter.java:9]:
```suggestion
    public String execute(String input) {
```

File: Лабораториска 1/PipeAndFilter/src/main/java/service/impl/ToLowerFilter.java [hunk 10]
```diff
@@ -0,0 +1,11 @@
+package service.impl;
+
+import service.Filter;
+
+public class ToLowerFilter implements Filter {
```
Comment (vasilaking) [inline Лабораториска 1/PipeAndFilter/src/main/java/service/impl/ToLowerFilter.java:5]:
```suggestion
public class ToLowerFilter<T> implements Filter<String> {
```

File: Лабораториска 1/PipeAndFilter/src/main/java/service/impl/ToLowerFilter.java [hunk 11]
```diff
@@ -0,0 +1,11 @@
...
+    @Override
+    public String execute(Object input) {
```
Comment (vasilaking) [inline Лабораториска 1/PipeAndFilter/src/main/java/service/impl/ToLowerFilter.java:7]:
```suggestion
    public String execute(String input) {
```

**Review by VasilakiG — CHANGES_REQUESTED** [review]
It is a good practice to specify the generics in the class definition.

"When using generics, it is a best practice to actually specify the generic type when creating an instance of the class. This seems to be the most object oriented solution."
-more about it here:
https://stackoverflow.com/questions/14586547/without-specifying-a-generic-type-method-does-not-return-a-list-of-elements-bu

File: Лабораториска 1/PipeAndFilter/src/main/java/PipeAndFilterProblem.java (hunk 3 above)
Comment (bdgToji) [reply Лабораториска 1/PipeAndFilter/src/main/java/PipeAndFilterProblem.java:63]: Forgot to change it while testing it, doesn't change anything if its in the inner if or the outer.

**Pull request merged**